include/chainbase/chainrocks.hpp

```diff
@@ -0,0 +1,818 @@
+#pragma once
+
+#include <boost/core/demangle.hpp>   // boost::core::demangle
+#include <boost/filesystem.hpp>      // boost::filesystem::path
+#include <boost/throw_exception.hpp> // BOOST_THROW_EXCEPTION
+
+#include <rocksdb/db.h>          // rocksdb::DB
+#include <rocksdb/write_batch.h> // rocksdb::WriteBatch
+
+#include <deque>    // std::deque
+#include <iostream> // std::cout
+#include <limits>   // std::numeric_limits
+#include <map>      // std::map
+#include <set>      // std::set
+#include <sstream>  // std::stringstream
+#include <vector>   // std::vector
+
+namespace chainrocks {
+
+   /// Holds the current undo state of a particular session.  For
+   /// example: whenever `start_undo_session` gets called, a new
+   /// `session` object is returned, and with that an `undo_state`
+   /// object gets pushed onto the `_undo_stack` signifying that
+   /// particular session with its own unique `_revision` number. In
+   /// turn, whenever the current `_state` is modified, these changes
+   /// get recorded in the `undo_state` object. Then if a user chooses
+   /// to undo whatever changes they've made, the information is
+   /// readily available to revert back to the original state of
+   /// `_state` safely due to this data structure.
+   struct undo_state {
+      undo_state()
+      {
+      }
+
+      /// Mapping to hold any modifications made to `_state`.
+      std::map<uint64_t, std::string> _modified_values{};
+
+      /// Mapping to hold any removed values from `_state`.
+      std::map<uint64_t, std::string> _removed_values{};
+
+      /// Set representing new keys that have been edded to `_state`.
+      std::set<uint64_t> _new_keys{};
+
+      /// The unique revision number held by each `undo_state` object.
+      /// Note that this revision number will never be less than or
```

Comment:
With the current implementation this comment is true.
But this will change in the future.

```
Process 75600 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = step over
    frame #0: 0x0000000100010e95 token_transfer_emulation_test`database_test::_initial_database_state(this=0x00007ffeefbfde40) at token_transfer_emulation_test.cpp:317:68
   314 	
   315 	   inline void _initial_database_state() {
   316 	      std::cout << "Filling initial database state... " << std::flush;
-> 317 	      for (size_t i{}; i < _gen_data.num_of_accounts_and_values(); ++i) {
    	                                                                   ^
   318 	         auto session{_database.start_undo_session(true)};
   319 	         _database.put(_gen_data.accounts()[i], std::to_string(_gen_data.values()[i]));
   320 	         session.push();
Target 0: (token_transfer_emulation_test) stopped.
(lldb) p *this
p *this
(database_test) $0 = {
  _database = {
    chainrocks::index = {
      _state = size=5 {
        [0] = (first = 8624466571468930473, second = "12005350491520422105")
        [1] = (first = 11187911585003888379, second = "11833621031671749371")
        [2] = (first = 12062998799301078992, second = "3773123098999608047")
        [3] = (first = 12723329799204884161, second = "15564293509712613380")
        [4] = (first = 17759844545201381884, second = "12793564622837496484")
      }
      _stack = size=5 {
        [0] = {
          _modified_values = size=0 {}
          _removed_values = size=0 {}
          _new_keys = size=1 {
            [0] = 12062998799301078992
          }
          _revision = 1
        }
        [1] = {
          _modified_values = size=0 {}
          _removed_values = size=0 {}
          _new_keys = size=1 {
            [0] = 12723329799204884161
          }
          _revision = 2
        }
        [2] = {
          _modified_values = size=0 {}
          _removed_values = size=0 {}
          _new_keys = size=1 {
            [0] = 17759844545201381884
          }
          _revision = 3
        }
        [3] = {
          _modified_values = size=0 {}
          _removed_values = size=0 {}
          _new_keys = size=1 {
            [0] = 8624466571468930473
          }
          _revision = 4
        }
        [4] = {
          _modified_values = size=0 {}
          _removed_values = size=0 {}
          _new_keys = size=1 {
            [0] = 11187911585003888379
          }
          _revision = 5
        }
      }
      _revision = 5
    }
    _database = { ... }
  }
  _gen_data = { ... }
  _log = { ... }
}
(lldb) 
```